### PR TITLE
Use public LifecycleObserver instead of internal

### PR DESCRIPTION
**Pull request by NotWoods**
Removed used of [internal GenericLifecycleObserver](https://github.com/mozilla-mobile/android-components/pull/3122#discussion_r287835670)

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt [hunk 1]
```diff
@@ -35,17 +39,18 @@ class ObserverRegistry<T> : Observable<T> {
     }
 
     override fun register(observer: T, owner: LifecycleOwner, autoPause: Boolean) {
-        if (owner.lifecycle.currentState == Lifecycle.State.DESTROYED) {
+        // Don't register if the owner is already destroyed
+        if (!owner.lifecycle.currentState.isAtLeast(CREATED)) {
             return
```
Comment (pocmo) [inline components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: Is `currentState` something we can't access anymore? This is still what `LiveData` does internally:
https://android.googlesource.com/platform/frameworks/support/+/androidx-master-dev/lifecycle/livedata-core/src/main/java/androidx/lifecycle/LiveData.java#190

Comment (NotWoods) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: We can still access it, but `isAtLeast` covers the `INITIALIZED` state too. If `LiveData` just uses equals internally I could switch back.

Comment (pocmo) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: I wonder if there can be a situation where we pass in a `LifecycleOwner` that is not CREATED yet and then we would never register? 🤔

Comment (NotWoods) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: [During onCreate](https://stackoverflow.com/questions/45459036/lifecycle-state-always-initialized), before the `ON_CREATE` event is triggered. However, it's probably more sensible to allow the observer to be registered during that time.

File: components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt [hunk 2]
```diff
@@ -35,17 +39,18 @@ class ObserverRegistry<T> : Observable<T> {
...
-        val lifecycleObserver = LifecycleBoundObserver(
-                owner,
-                registry = this,
-                observer = observer,
-                autoPause = autoPause)
+        val lifecycleObserver = if (autoPause) {
+            AutoPauseLifecycleBoundObserver(owner, registry = this, observer = observer)
+        } else {
+            LifecycleBoundObserver(owner, registry = this, observer = observer)
+        }
```
Comment (pocmo) [inline components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: If we specify `autoPause` then we do not have the destroy behavior here anymore, or am I missing something?

Comment (NotWoods) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: `AutoPauseLifecycleBoundObserver` extends `LifecycleBoundObserver`

Comment (pocmo) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: Oops, missed that, somehow I thought it extended `LifecycleObserver`.

**Pull request merged**